src/native-desktop/RootView.tsx

```diff
@@ -0,0 +1,165 @@
+            // focusable elements.
+            this._focusManager = new FocusManager(undefined);
+        }
+
+        _onTouchStartCapture = (e: SyntheticEvent) => {
+            this._updateKeyboardNavigationState(false);
+        }
+
+        _onKeyDownCapture = (e: SyntheticEvent) => {
+
+            let kbdEvent = EventHelpers.toKeyboardEvent(e);
+            if (kbdEvent.keyCode === KEY_CODE_TAB) {
+                this._updateKeyboardNavigationState(true);
+            }
+
+            if (kbdEvent.keyCode === KEY_CODE_ESC) {
+                // If Esc is pressed and the focused element stays the same after some time,
+                // switch the keyboard navigation off to dismiss the outline.
+                const activeComponent = FocusManager.getCurrentFocusedComponent();
+
+                if (this._isNavigatingWithKeyboardUpateTimer) {
+                    clearTimeout(this._isNavigatingWithKeyboardUpateTimer);
+                }
+
+                this._isNavigatingWithKeyboardUpateTimer = setTimeout(() => {
+                    this._isNavigatingWithKeyboardUpateTimer = undefined;
+
+                    if (activeComponent === FocusManager.getCurrentFocusedComponent()) {
+                        this._updateKeyboardNavigationState(false);
+                    }
+                }, 500);
+            }
+        }
+
+        _updateKeyboardNavigationState(isNavigatingWithKeyboard: boolean) {
+            if (this._isNavigatingWithKeyboardUpateTimer) {
+                window.clearTimeout(this._isNavigatingWithKeyboardUpateTimer);
+                this._isNavigatingWithKeyboardUpateTimer = undefined;
+            }
+
+            if (this._isNavigatingWithKeyboard !== isNavigatingWithKeyboard) {
+                this._isNavigatingWithKeyboard = isNavigatingWithKeyboard;
+
+                UserInterface.keyboardNavigationEvent.fire(isNavigatingWithKeyboard);
+            }
+        }
+
+        _onKeyDown = (e: SyntheticEvent) => {
+            let kbdEvent = EventHelpers.toKeyboardEvent(e);
+            Input.dispatchKeyDown(kbdEvent);
+        }
+
+        _onKeyUp = (e: SyntheticEvent) => {
+            let kbdEvent = EventHelpers.toKeyboardEvent(e);
+
+            let activePopupId = FrontLayerViewManager.getActivePopupId();
+            if (activePopupId && (kbdEvent.keyCode === KEY_CODE_ESC)) {
+                if (e.stopPropagation) {
+                    e.stopPropagation();
+                }
+                FrontLayerViewManager.dismissPopup(activePopupId);
+                return;
+            }
+
+            Input.dispatchKeyUp(kbdEvent);
+        }
+
+        getChildContext() {
+            // Provide the context with root FocusManager to all descendants.
+            return {
+                focusManager: this._focusManager
+            };
+        }
+
+        render() {
+            let content = super.render();
+
+            // Using "any" since onKeyDown/onKeyDown are not defined at RN.View property level
```

Comment:
> onKeyDown/onKeyDown [](start = 32, length = 20)

This says `onKeyDown` twice

Reply:
lol

---
In reply to: [159357269](https://github.com/Microsoft/reactxp/pull/404#discussion_r159357269) [](ancestors = 159357269)